src/EFCore.PG/Query/Internal/NpgsqlEvaluatableExpressionFilter.cs

```diff
@@ -11,11 +13,23 @@ public class NpgsqlEvaluatableExpressionFilter : RelationalEvaluatableExpression
             nameof(NpgsqlDbFunctionsExtensions.Cast),
             BindingFlags.Public | BindingFlags.Static);
 
+        private static readonly MethodInfo _tsQueryParse = typeof(NpgsqlTsQuery).GetMethod(
+            nameof(NpgsqlTsQuery.Parse),
+            BindingFlags.Public | BindingFlags.Static);
+
+        private static readonly MethodInfo _tsVectorParse = typeof(NpgsqlTsVector).GetMethod(
+            nameof(NpgsqlTsVector.Parse),
+            BindingFlags.Public | BindingFlags.Static);
+
         public NpgsqlEvaluatableExpressionFilter([NotNull] IModel model) : base(model) { }
 
         public override bool IsEvaluatableMethodCall(MethodCallExpression methodCallExpression) =>
             (!methodCallExpression.Method.IsGenericMethod
              || methodCallExpression.Method.GetGenericMethodDefinition() != _cast)
+            && methodCallExpression.Method != _tsQueryParse
```

Comment:
This is because EF Core tries to compile the LINQ query and does a first pass optimization using ParameterExtractingExpressionVisitor eliminating expressions that do not reference any columns and evaluating them on the client side and sending the results directly as a literal.

Without this my unit tests break and it introduces a subtle behavior anomaly where if you do something like this:

```
context.Query<SomeEntity>(x => 1 + 1);
```
The 1+1 expression will be evaluated client-side and not translated to SQL.

But if you do this:
```
context.Query<SomeEntity>(x => x + 1);
```
It will be properly be translated to SQL.

Entity Framework Core people were aware of this issue from before and had used a workaround for this (FunctionEvaluationDisablingVisitor) to avoid re-linq doing this optimization when not desired, but later on a native way was introduced in re-linq (EvaluatableExpressionFilter) so they removed that visitor and used the new way instead.

Note this affects your Regex implementation as well, you may want to add Regex methods to the filter.

Reply:
Thanks for explaining this. The thing still causing me trouble, is why this isn't happening in EF Core, the SQL Server provider or anywhere else I can see.

EF Core's [base EvaluatableExpressionFilter](https://github.com/aspnet/EntityFrameworkCore/blob/dev/src/EFCore/Query/Internal/EvaluatableExpressionFilter.cs) only lists a handful of non-deterministic methods such as `DateTime.Now`, while [RelationalEvaluatableExpressionFilter](https://github.com/aspnet/EntityFrameworkCore/blob/dev/src/EFCore.Relational/Query/Internal/RelationalEvaluatableExpressionFilter.cs) adds user-defined database functions to that. The SQL Server and Sqlite providers don't have a EvaluatableExpressionFilter at all.

In other words, the entire rich array of methods that EF Core knows how to translate (including SQL Server's `FREETEXT` method (full-text search) is absent from EvaluatableExpressionFilter. Any idea why that is? We can also ask the EF Core team about this.

Regarding the Regex implementation, isn't it actually an advantage that an expression gets evaluated locally when possible? I mean, if the user writes an expression with regex that *doesn't* reference any database columns (this would be weird), why should that get sent to the database for evaluation?

Reply:
@smitpatel, @anpete there are some good query design questions here that I don't feel prepared to answer accurately... Can you weigh in? 

In general I do agree with @roji that we prefer evaluating things that are uncorrelated to the server on the client.

Also, not sure you care but this is probably not a safe way to compare reflection objects across all .NET implementations.

Reply:
> why this isn't happening in EF Core, the SQL Server provider or anywhere else I can see.

Because, I think, these new functions are unusual in that they may be used in a way where they are not correlated to any range variable in the query (at least that's what it looks like from looking at the tests). In the more normal case, the function call expression contains one or more range variable expressions, and so would never be considered for partial eval.

We may want to allow providers to simply add entries to the base EvaluatableExpressionFilter in order to avoid a service replacement here..

Reply:
@anpete I don't think the these new functions are very special - it's true that the tests currently seem to exercise mostly local data, but the functions obviously make sense mostly with database data (after all we're talking about full-text search)...

Regardless, definitely a +1 for allowing adding entries in EvaluatableExpresisonFilter (like the member/method translators?) - and so also moving it out of Internal.